File: lib/decoders/fluxdecoder.cc
#include "lib/core/globals.h"
#include "lib/data/fluxmap.h"
#include "lib/data/fluxmapreader.h"
#include "lib/decoders/fluxdecoder.h"
#include "lib/decoders/decoders.pb.h"

/* This is a port of the samdisk code:
 *
 * https://github.com/simonowen/samdisk/blob/master/src/FluxDecoder.cpp
 *
 * I'm not actually terribly sure how it works, but it does, and much better
 * than my code.
 */

FluxDecoder::FluxDecoder(
    FluxmapReader* fmr, nanoseconds_t bitcell, const DecoderProto& config):
    _fmr(fmr),
    _pll_phase(config.pll_phase()),
    _pll_adjust(config.pll_adjust()),
    _flux_scale(config.flux_scale()),
    _clock(bitcell),
    _clock_centre(bitcell),
    _clock_min(bitcell * (1.0 - _pll_adjust)),
    _clock_max(bitcell * (1.0 + _pll_adjust)),
    _flux(0),
    _leading_zeroes(fmr->tell().zeroes)
{
}

bool FluxDecoder::readBit()
{
    if (_leading_zeroes > 0)
    {
        _leading_zeroes--;
        return false;
    }
    else if (_leading_zeroes == 0)
    {
        _leading_zeroes--;
        return true;
    }

    while (!_fmr->eof() && (_flux < (_clock / 2)))
    {
        _flux += nextFlux() * _flux_scale;
        ;
        _clocked_zeroes = 0;
    }

    _flux -= _clock;
    if (_flux >= (_clock / 2))
    {
        _clocked_zeroes++;
        _goodbits++;
        return false;
    }

    /* PLL adjustment: change the clock frequency according to the phase
     * mismatch */
    if (_clocked_zeroes <= 3)
    {
        /* In sync: adjust base clock */

        _clock += _flux * _pll_adjust;
    }
    else
    {
        /* Out of sync: adjust the base clock back towards the centre */

        _clock += (_clock_centre - _clock) * _pll_adjust;

        /* We require 256 good bits before reporting another sync loss event. */

        if (_goodbits >= 256)
            _sync_lost = true;
        _goodbits = 0;
    }

    /* Clamp the clock's adjustment range. */

    _clock = std::min(std::max(_clock_min, _clock), _clock_max);

    /* I'm not sure what this does, but the original comment is:
     * Authentic PLL: Do not snap the timing window to each flux transition
     */

    _flux = _flux * (1.0 - _pll_phase);

    _goodbits++;
    return true;
}

std::vector<bool> FluxDecoder::readBits(unsigned count)
{
    std::vector<bool> result;
    while (!_fmr->eof() && count--)
    {
        bool b = readBit();
        result.push_back(b);
    }
    return result;
}

std::vector<bool> FluxDecoder::readBits(const Fluxmap::Position& until)
{
    std::vector<bool> result;
    while (!_fmr->eof() && (_fmr->tell().bytes < until.bytes))
    {
        bool b = readBit();
        result.push_back(b);
    }
    return result;
}

nanoseconds_t FluxDecoder::nextFlux()
{
    return _fmr->readInterval(_clock_centre) * NS_PER_TICK;
}

File: java/com/cowlark/fluxengine/decoders/FluxDecoder.java
package com.cowlark.fluxengine.decoders;

import static com.cowlark.fluxengine.external.FluxEngine.NS_PER_TICK;

import com.cowlark.fluxengine.core.Bits;
import com.cowlark.fluxengine.data.FluxPosition;
import com.cowlark.fluxengine.data.FluxmapReader;

/* This is a port of the samdisk code:
 *
 * https://github.com/simonowen/samdisk/blob/master/src/FluxDecoder.cpp
 *
 * I'm not actually terribly sure how it works, but it does, and much better
 * than my code.
 */
public class FluxDecoder
{
    private final FluxmapReader fmr;
    private final double pllPhase;
    private final double pllAdjust;
    private final double fluxScale;
    private final double clockCentreNs;
    private final double clockMinNs;
    private final double clockMaxNs;
    private double clockNs;
    private double fluxNs = 0.0;
    private int clockedZeroes = 0;
    private int goodbits = 0;
    private boolean index = false;
    private boolean syncLost = false;
    private int leadingZeroes;

    public FluxDecoder(FluxmapReader fmr, double bitcellNs, DecoderProto config)
    {
        this.fmr = fmr;
        pllPhase = config.getPllPhase();
        pllAdjust = config.getPllAdjust();
        fluxScale = config.getFluxScale();
        clockNs = bitcellNs;
        clockCentreNs = bitcellNs;
        clockMinNs = bitcellNs * (1.0 - pllAdjust);
        clockMaxNs = bitcellNs * (1.0 + pllAdjust);
        leadingZeroes = fmr.tell().zeroes();
    }

    private static double clampClock(double min, double value, double max)
    {
        if (value > max)
            return max;
        if (value < min)
            return min;
        return value;
    }

    public boolean readBit()
    {
        if (leadingZeroes > 0)
        {
            leadingZeroes--;
            return false;
        } else if (leadingZeroes == 0)
        {
            leadingZeroes--;
            return true;
        }

        while (!fmr.eof() && fluxNs < clockNs / 2.0)
        {
            fluxNs += nextFlux() * fluxScale;
            clockedZeroes = 0;
        }

        fluxNs -= clockNs;
        if (fluxNs >= clockNs / 2.0)
        {
            clockedZeroes++;
            goodbits++;
            return false;
        }

        /* PLL adjustment: change the clock frequency according to the phase
         * mismatch */
        if (clockedZeroes <= 3)
        {
            /* In sync: adjust base clock */

            clockNs += fluxNs * pllAdjust;
        } else
        {
            /* Out of sync: adjust the base clock back towards the centre */

            clockNs += (clockCentreNs - clockNs) * pllAdjust;

            /* We require 256 good bits before reporting another sync loss
             * event. */

            if (goodbits >= 256)
                syncLost = true;
            goodbits = 0;
        }

        /* Clamp the clock's adjustment range. */

        clockNs = clampClock(clockMinNs, clockNs, clockMaxNs);

        /* I'm not sure what this does, but the original comment is:
         * Authentic PLL: Do not snap the timing window to each flux
         * transition */

        fluxNs *= 1.0 - pllPhase;

        goodbits++;
        return true;
    }

    public Bits readBits(int count)
    {
        Bits result = new Bits();
        while (!fmr.eof() && count-- > 0)
            result.add(readBit());
        return result;
    }

    public Bits readBits(FluxPosition until)
    {
        Bits result = new Bits();
        while (!fmr.eof() && fmr.tell().bytes() < until.bytes())
            result.add(readBit());
        return result;
    }

    public Bits readBits()
    {
        return readBits(Integer.MAX_VALUE);
    }

    private double nextFlux()
    {
        long ticks = fmr.readInterval((long) (clockCentreNs / NS_PER_TICK));
        return ticks * NS_PER_TICK;
    }
}